tests/Graphics/Implicit/Test/Instances.hs

```diff
@@ -0,0 +1,234 @@
+{-# LANGUAGE FlexibleInstances     #-}
+{-# LANGUAGE MultiParamTypeClasses #-}
+{-# LANGUAGE TypeApplications      #-}
+
+{-# OPTIONS_GHC -fno-warn-orphans #-}
+
+module Graphics.Implicit.Test.Instances (Quantizable (quantize), epsilon, observe, (=~=)) where
+
+import Data.VectorSpace (AdditiveGroup((^-^)))
+import qualified Graphics.Implicit as I
+import Graphics.Implicit
+    ( ExtrudeRMScale(Fn, C1, C2),
+      SymbolicObj3,
+      SymbolicObj2,
+      ℝ3,
+      ℝ2,
+      ℝ,
+      squareR,
+      Object(shell, translate, unionR, intersectR, differenceR),
+      sphere,
+      cubeR,
+      cylinder2,
+      cylinder,
+      circle,
+      polygonR,
+      extrudeR,
+      rotate3,
+      rotate3V,
+      rotate )
+import Graphics.Implicit.Definitions
+    ( SymbolicObj3(Cylinder, Complement3, UnionR3, DifferenceR3,
+                   IntersectR3, Translate3, Scale3, Rotate3, Rotate3V, Outset3,
+                   Shell3, CubeR),
+      SymbolicObj2(SquareR, Complement2, UnionR2, DifferenceR2,
+                   IntersectR2, Translate2, Scale2, Rotate2, Outset2, Shell2,
+                   PolygonR),
+      both,
+      allthree )
+import Graphics.Implicit.Primitives ( Object(getBox, getImplicit) )
+import Prelude (Bool (True, False), Int, Double, Integer, (.), flip, uncurry, any, ($), (==), not, (>), (<), and, (&&), all, (>=), length, div, (<*>), (<$>), (+), fmap, (/), fromIntegral, (^), (*), (<>), round, (<=), filter )
+import QuickSpec ( Observe(observe), (=~=) )
+import Test.QuickCheck
+    ( Arbitrary(arbitrary, shrink),
+      genericShrink,
+      choose,
+      oneof,
+      scale,
+      sized,
+      vectorOf,
+      Gen,
+      Positive(getPositive) )
+
+
+------------------------------------------------------------------------------
+-- | The number of decimal points we need to agree to assume two 'Double's are
+-- equal.
+epsilon :: Int
```

Comment:
because of the nature of floating point, anything that looks at decimal position is going to be fragile. there's functions for taking the 'integer' part and the 'multiplier' part and examining those seperately, for cases like this.

Reply:
https://hackage.haskell.org/package/ieee-0.7/docs/Numeric-IEEE.html

take a look at sameSignificandBits . i think. I'm good with merging this without this, it seems good enough already.

Reply:
I'd prefer to merge this as is, but wouldn't push back too strongly if you're keen on it.